Add combinators to ConfigReader, ConfigWriter and ConfigConvert

ruippeixotog opened the pull request:
Adds some combinators to config reading/writing classes so they are easier to compose.

I ended up not needing anything from cats; my previous concerns with F-bounded polymorphism didn't apply to this specific situation. Now that we have these methods defined we should probably create a `pureconfig-cats` module (and a `pureconfig-scalaz`, if anybody asks for it) with type classes for our three classes.

Closes #93.

ruippeixotog committed:
Change the type parameter names of config type classes

ruippeixotog committed:
Add combinators to config reading/writing classes

ruippeixotog commented on core/src/test/scala/pureconfig/ConfigReaderSuite.scala:
```diff
@@ -0,0 +1,70 @@
+    def zip[A, B](r1: ConfigReader[A], r2: ConfigReader[B]): Either[ConfigReaderFailures, (A, B)] = {
+      (r1.from(conf), r2.from(conf)) match {
+        case (Right(a), Right(b)) => Right((a, b))
+        case (Left(fa), _) => Left(fa)
```
Right now, `zip` short-circuits in case the first reader fails. We could also use always both readers and combine their failures if both fail. What do you prefer?

melrief replied:
I prefer to run boths and combine their failures as we do for products in general. The only advantage of short-circuiting is that it runs faster but for us speed is not an issue.

melrief reviewed (approved):
Great PR, thanks!

I have two general comments:
1. you have added the `fromFunction` functions to `ConfigReader` and `ConfigWriter` to create new instances from a provided function, I would consider adding a `fromFunctions` to `ConfigConvert` to do the same.
2. I've noticed that you didn't add `ConfigReader.emap` to map over the underlying result type. I propose to add it as simplified version of `flatMap`, something like:

```scala
def emap[B](f: A => Either[ConfigReaderFailures, B]): ConfigReader[B] =
   fromFunction[B] { cv: ConfigValue => from(cv).right.flatMap(f) }
```

Both points are optional and I approve the PR.

ruippeixotog committed:
Add an emap method to ConfigReader

melrief commented:
Thanks for `emap`, LGTM

ruippeixotog committed:
Make ConfigReader#zip show errors from both sides on failure

ruippeixotog committed:
Fix specs and increase number of scalacheck iterations

ruippeixotog commented:
@melrief I have just changed the behavior of `zip` in 02e5bda, I agree that it's better to show errors from both sides.

I also did some minor changes in the specs in 29ecd77. It seems that `===` doesn't behave in the way I thought (I'm a specs2 user), so I replaced it with `shouldEqual`. I also increased the number of scalacheck iterations to 100 in those files, as I think 10 is not enough for most purposes.

jcazevedo reviewed (approved):
LGTM!

Thanks for this and also for changing the name of the type parameters from the Java style to the Scala style. We should strive to replace the remaining usages of the Java style and to make the usage of the Scala style consistent in future PRs.

melrief committed:
Merge branch 'master' into reader-writer-combinators
